Implementation of torus surfaces

paulromano opened the pull request:
This PR adds the long awaited `XTorus`, `YTorus`, and `ZTorus` surfaces to OpenMC. The implementation was started by @makeclean who got it 90% of the way there so most of the credit should go to him. I picked up his branch and made some modest changes to get it working. Importantly, instead of relying on our own quartic (4th order polynomial) root finding algorithm, I picked up a solver that was recently published in [ACM TOMS](https://doi.org/10.1145/3386241) by Orellana and De Michele. The paper has an accompanying C implementation which I've minimally translated to C++. No license was present in the C implementation to that paper but I reached out to the corresponding author (@cridemichel) and he has agreed to license it under the BSD license (included in the PR). Documentation and unit/regressions tests have been updated accordingly.

shimwell commented on tests/regression_tests/torus/test.py:
```diff
@@ -0,0 +1,39 @@
+    al.add_nuclide('H1', 1.0)
+    model.materials.extend([fuel, al])
+
+    # 🍩🍩🍩
```
nice use of doughnuts

paulromano replied:
That _is_ why unicode was invented, right? So we can put donuts in our code :smile:

gridley replied:
Wow :open_mouth: I guess you could say unicode's applications have come... full circle

gridley commented:
This is really neat. I wonder what shapes could be done with a general quartic surface.

makeclean commented:
![plot_1](https://user-images.githubusercontent.com/2439754/147288917-7919af20-28e1-4219-8ce2-1f0c4a6aeef2.png)
Just to demonstrate it in action :)

pshriwise commented on docs/source/methods/geometry.rst:
```diff
@@ -899,6 +977,27 @@ Dxy + Eyz + Fxz + Gx + Hy + Jz + K = 0`. Thus, the gradient to the surface is
+Torus Parallel to an Axis
+-------------------------
+
+A cone parallel to, for example, the x-axis has the form
```
```suggestion
A torus parallel to, for example, the x-axis has the form
```

pshriwise commented on include/openmc/surface.h:
```diff
@@ -371,6 +371,57 @@ class SurfaceQuadric : public CSGSurface {
+//==============================================================================
+//! A toroidal surface described by the quartic torus lies in the y direction
+//
+//! \f$(x-x_0)^2 + (\sqrt{(y-y_0)^2 + (z-z_0)^2} - A)^2/C^2 -1 \f$
```
```suggestion
//! \f$(y-y_0)^2/B^2 + (\sqrt{(x-x_0)^2 + (z-z_0)^2} - A)^2/C^2 -1 \f$
```

pshriwise commented on include/openmc/surface.h:
```diff
@@ -371,6 +371,57 @@ class SurfaceQuadric : public CSGSurface {
+//==============================================================================
+//! A toroidal surface described by the quartic torus lies in the z direction
+//
+//! \f$(x-x_0)^2 + (\sqrt{(y-y_0)^2 + (z-z_0)^2} - A)^2/C^2 -1 \f$
```
```suggestion
//! \f$(z-z_0)^2/B^2 + (\sqrt{(x-x_0)^2 + (y-y_0)^2} - A)^2/C^2 -1 \f$
```

pshriwise commented on include/openmc/surface.h:
```diff
@@ -371,6 +371,57 @@ class SurfaceQuadric : public CSGSurface {
+//==============================================================================
+//! A toroidal surface described by the quartic torus lies in the x direction
+//
+//! \f$(x-x_0)^2 + (\sqrt{(y-y_0)^2 + (z-z_0)^2} - A)^2/C^2 -1 \f$
```
```suggestion
//! \f$(x-x_0)^2/B^2 + (\sqrt{(y-y_0)^2 + (z-z_0)^2} - A)^2/C^2 -1 \f$
```

pshriwise commented on src/surface.cpp:
```diff
@@ -85,6 +87,25 @@ void read_coeffs(pugi::xml_node surf_node, int surf_id, double& c1, double& c2,
+  int stat = sscanf(
+    coeffs.c_str(), "%lf %lf %lf %lf %lf %lf", &c1, &c2, &c3, &c4, &c5, &c6);
+  if (stat != 6) {
+    fatal_error("Something went wrong reading surface coeffs");
```
Since we have the info here, we may as well report the surface ID in the error message too.

pshriwise commented on src/surface.cpp:
```diff
@@ -983,6 +1004,212 @@ void SurfaceQuadric::to_hdf5_inner(hid_t group_id) const
+  // zero but possibly small and positive. A tolerance is set to discard that
+  // zero.
+  double distance = INFTY;
+  double cutoff = coincident ? 1e-9 : 0.0;
```
Would one of the already defined constants (i.e. `FP_COINCIDENT`) work here or is `1e-9` a sweet spot for this check?

paulromano replied:
Unfortunately not. I was running into cases where I'd get a root greater than `FP_COINCIDENT` when the particle was coincident. A lower cutoff would probably still work (e.g., 1e-10), but I might have a better solution anyway in the coincident case, which is to just solve a cubic instead since we know one root should be zero.

paulromano replied:
An update here: I did try out the approach of solving a cubic equation when the particle is coincident with the surface, but it didn't seem to yield any measurable increase in performance. Given that it would require adding more code, I'm just going to stick with our current approach, which is to solve the quartic anyway and skip the expected root near zero.

As far as the value of the cutoff, I ran the regression test model with a few million neutrons and the highest value that I got for the expected zero root was about 1.3e-11, so it's probably safe to reduce the cutoff here to 1e-10. Let me know how you'd like me to proceed.

pshriwise replied:
> An update here: I did try out the approach of solving a cubic equation when the particle is coincident with the surface, but it didn't seem to yield any measurable increase in performance. Given that it would require adding more code, I'm just going to stick with our current approach, which is to solve the quartic anyway and skip the expected root near zero.

Thanks for testing that out! 

> As far as the value of the cutoff, I ran the regression test model with a few million neutrons and the highest value that I got for the expected zero root was about 1.3e-11, so it's probably safe to reduce the cutoff here to 1e-10. Let me know how you'd like me to proceed.

I'd say yeah let's reduce it to `1e-10` then. Thanks for pinning that down. Normally 
I'd request a new constant for clarity, but your comment right above is pretty thorough.

pshriwise commented on src/surface.cpp:
```diff
@@ -983,6 +1004,212 @@ void SurfaceQuadric::to_hdf5_inner(hid_t group_id) const
+  write_dataset(group_id, "coefficients", coeffs);
+}
+
+// todo should do some alebgra first to get rid of the sqrt
```
Could we place this comment closer to the relevant `sqrt`?

paulromano replied:
I'm just going to remove this comment -- on modern architectures `sqrt` is actually about as expensive as a `div` instruction, so probably not something we need to be overly worried about

pshriwise commented on src/surface.cpp:
```diff
@@ -983,6 +1004,212 @@ void SurfaceQuadric::to_hdf5_inner(hid_t group_id) const
+  double z = r.z - z0_;
+
+  // f(x,y,z) = x^2/B^2 + (sqrt(y^2 + z^2) - A)^2/C^2 - 1
+  // ∂f/∂x = 2x/B^2
```
Fancy comments! 💯

pshriwise reviewed (changes requested):
This is great @paulromano! Thank you to you and @makeclean for this capability. Mostly documentation comments from me. 

I'm going to play around with it a bit in the next day or two to try out a few edge cases we might run into.

paulromano commented on docs/source/methods/geometry.rst:
```diff
@@ -437,6 +449,72 @@ Defining the terms
+.. math::
+    :label: dist-xtorus-4
+
+    \begin{align*}
```
Strange, it comes up fine on mine:
![image](https://user-images.githubusercontent.com/743095/147676015-8f4b4fa3-c779-4186-8662-528711c98596.png)

pshriwise replied:
Yep! False alarm. I was using an older version of sphinx with bug for multi-line equations that has since been fixed.

paulromano commented:
@pshriwise I just made a few updates, addressing most of your comments and fixing a bug (in the [quartic solver](https://github.com/cridemichel/quarticpp/issues/1)) that was causing lost particles when the major radius was close to the minor radius. I also just added a `rotate` method for our torus Python classes that can handle "simple" rotations (anything with multiples of 90 degrees).

pshriwise reviewed (approved):
Thanks for pushing this to the finish line @paulromano! And thanks @makeclean for getting it most of the way there.

pshriwise commented:
Noting that there's an additional fix for CI in https://github.com/openmc-dev/openmc/pull/1933/commits/67db1f5af0c32d83a0f40595132b6d4a3dac903c. The release of Numpy 1.22.0 has caused some problems with floating point consistency when writing XML files on GHA.
